File: src/leetcode/trees/cpp/104.cpp
#include "eval.cpp"
#include "stddef.h"
#include <cstdio>
#include <sys/param.h>
#include <vector>

using namespace std;

int recursive_max_depth_binary_tree(Node *root) {
  if (root != nullptr) {
    int leftDepth = recursive_max_depth_binary_tree(root->left);
    int rightDepth = recursive_max_depth_binary_tree(root->right);

    return MAX(leftDepth, rightDepth) + 1;
  }

  return 0;
}

int main(int argc, char *argv[]) {
  // Parse arguments
  vector<int> input = parse_single_array(argv[1]);
  int expected_output = atoi(argv[2]);

  // Create Trees and get depth
  Node *inputTree = constructor(input);
  int output = recursive_max_depth_binary_tree(inputTree);

  if (output != expected_output) {
    printf(
        "Error getting depth(104) for provided input tree expected %d got %d\n",
        expected_output, output);
    return 1;
  }

  return 0;
}

File: src/leetcode/trees/python/104.py
#!/usr/bin/env python

from typing import Optional

__all__ = ["recursive_max_depth_binary_tree"]


class Node:
    def __init__(self, val=0, left=None, right=None) -> None:
        self.val = val
        self.left = left
        self.right = right


def recursive_max_depth_binary_tree(root: Optional[Node]) -> int:
    if root:
        return (
            max(
                recursive_max_depth_binary_tree(root.left),
                recursive_max_depth_binary_tree(root.right),
            )
            + 1
        )
    else:
        return 0
